File: hacker_rank/interview_street/NonTimedChallenges/DynamicProgramming/FairyChess/FairyChess.cxx
#include <cstdio>
#include <cstring>
#include <algorithm>
#include <iostream>

using namespace std;

#define mod 1000000007
#define REP(i, n) for(int i = 0; i < (n); ++i)
#define FOR(i, st, nd) for(int i = (st); i < (nd); ++i)

typedef long long int LL;

// handles the negative numbers
inline LL modme(LL num) { return num < 0 ? mod + num : num%mod; }

char inputBoard[200][200];
LL origBoard[400][400];
LL sumBoard[400][400];
LL memo[400][400];
LL startIndex[400];
LL endIndex[400];
int N, S, M, newN; // newN will be 2*N-1
int Lx, Ly;

template<typename T>
void printBoard(T &board, int size) {
	REP(i, size) {
		REP(j, size) cout << board[i][j] << "\t";
		printf("\n");
	}
}

// From the original input, this function creates a diamond shaped board, which 
// is double the size.
void initResultBoard() {
	newN = 2*N-1;
	memset(origBoard, 0, sizeof origBoard);
	memset(memo, 0, sizeof memo);
	memset(sumBoard, 0, sizeof sumBoard);
    
    // index in diamond
    int coldmndi = N-1, coldmndj = N;
    int stepcoli = -1, stepcolj = 1;
    int starti = 0, startj = 0;
    int incri = 1, incrj = 0;
    int row = 0, col = 0;
	REP(x, newN) {
		startIndex[x] = coldmndi;
		endIndex[x] = coldmndj;
		for(int y = coldmndi; y < coldmndj; y+=2) {

            if(inputBoard[row][col] == '.')
                origBoard[x][y] = 1;
			else if(inputBoard[row][col] == 'L') 
                origBoard[x][y] = 1, Lx = x, Ly = y;
			memo[x][y] = origBoard[x][y];

            row = row-1, col = col+1;
		}
        if (coldmndi <= 0) {
            swap(stepcoli, stepcolj);
			starti = N-1, startj = 0;
            incri = 0, incrj = 1;
		}
        coldmndi += stepcoli;
        coldmndj += stepcolj;
        starti += incri, startj += incrj;
        row = starti, col = startj;
	}
}
    
// Calculates the partial sum
void constructSumBoard() {
	int startj;
	REP(i, newN) {
		startj = i >= N ? 0 : startIndex[i];

		for(int j = startj; j < newN; ++j) {
			sumBoard[i][j] = memo[i][j];
            if(j > 0) sumBoard[i][j] += sumBoard[i][j-1];
            if(i > 0) sumBoard[i][j] += sumBoard[i-1][j];
			if(j > 0 && i > 0) 
				sumBoard[i][j] -= sumBoard[i-1][j-1];
			sumBoard[i][j] = modme(sumBoard[i][j]);
		}

	}
}

// Recalculate the board values based on the partial sums
void recalculateResultBoard() {
	int x_s_1, y_s_1, xs, ys;
	LL sub;
	REP(x, newN) {
        for(int y = startIndex[x]; y < endIndex[x]; y+=2) 
            if(origBoard[x][y] == 1) {
				xs = min(newN-1,x+S);
				ys = min(newN-1, y+S);
				x_s_1 = x-S-1, y_s_1 = y-S-1;
				sub = 0;
                memo[x][y] = sumBoard[xs][ys];
                if(x_s_1 >= 0 && y_s_1 >= 0)
                    memo[x][y] += sumBoard[x_s_1][y_s_1];
                if(x_s_1 >= 0) 
                    sub += sumBoard[x_s_1][ys];
                if(y_s_1 >= 0)
                    sub += sumBoard[xs][y_s_1];
				memo[x][y] = modme(memo[x][y]-sub);
			}
	}
}

// This is DP solution. The idea is to calculate the number of ways to move from
// the every location at each step. For step 1 it would be the total number of
// positions that can be reached from that location given S. For step 2 it would
// be the sum of these values at every location and so on.
// To do this easily for M steps, at every step we need to calculate the partial
// sum just like we do for the classic maxsum DP problem. Using these values, we
// can calculate the total number of ways at each step in O(N^2) ways. This gives
// us the total complexity at O(MN^2).
// Note that to do this partial sums easily, the board must be rotated to a 
// diamond shape so that our calculation is easy.
LL fairyChess() {
	initResultBoard();
	M--;
	REP(step, M) {
        constructSumBoard();
        recalculateResultBoard();
	}
	constructSumBoard();
	int x = Lx, y = Ly;
	int xs = min(newN-1,x+S);
	int ys = min(newN-1, y+S);
	int x_s_1 = x-S-1, y_s_1 = y-S-1;
	memo[x][y] = sumBoard[xs][ys];
	if(x_s_1 >= 0 && y_s_1 >= 0)
		memo[x][y] = (memo[x][y] + sumBoard[x_s_1][y_s_1])%mod;
	if(x_s_1 >= 0) 
		memo[x][y] = modme(memo[x][y] - sumBoard[x_s_1][ys]);
	if(y_s_1 >= 0)
		memo[x][y] = modme(memo[x][y] - sumBoard[xs][y_s_1]);

	
	return memo[Lx][Ly];
}
    
int main(int argc, char *argv[]) {
    int T; scanf(" %d", &T);
	while(T--) {
		scanf(" %d %d %d", &N, &M, &S);
		REP(i, N) 
			scanf(" %s", inputBoard[i]);

        printf("%lld\n", fairyChess());
	}
	return 0;
}

File: hacker_rank/interview_street/NonTimedChallenges/DynamicProgramming/FairyChess/FairyChess.py
import sys
import copy

mod = 1000000007

# From the original input, this function creates a diamond shaped board, which 
# is double the size.
def initResultBoard(board):
    """Returns the rotated board with integer values in it.

    >>> initResultBoard([".....","...L.", ".P.P.", "P.P..", "..P.."])
    ([[0, 0, 0, 0, 1, 0, 0, 0, 0], [0, 0, 0, 1, 0, 1, 0, 0, 0], [0, 0, 1, 0, 1, 0, 1, 0, 0], [0, 0, 0, 0, 0, 1, 0, 1, 0], [1, 0, 1, 0, 1, 0, 1, 0, 1], [0, 1, 0, 0, 0, 0, 0, 1, 0], [0, 0, 0, 0, 1, 0, 1, 0, 0], [0, 0, 0, 1, 0, 1, 0, 0, 0], [0, 0, 0, 0, 1, 0, 0, 0, 0]], 4, 6)
    """
    resBoard = []
    size = len(board)
    newsize = size*2 - 1
    midway = newsize/2
    # index in diamond
    coldmndi, coldmndj = midway, midway+1
    stepcoli, stepcolj = -1, 1
    starti, startj = 0, 0
    incri, incrj = 1, 0
    row, col = 0, 0
    lastIndex = []
    for x in xrange(newsize):
        lastIndex.append(coldmndj)
        resrow = [0]*newsize
        for y in xrange(coldmndi, coldmndj, 2):

            if board[row][col] == '.':
                resrow[y] = 1
            elif board[row][col] == 'L':
                resrow[y] = 1
                lx, ly = x, y
            row, col = row-1, col+1
        resBoard.append(resrow)
        if coldmndi <= 0:
            stepcoli, stepcolj = (stepcolj, stepcoli)
            starti, startj = (size-1, 0)
            incri, incrj = (0, 1)
        coldmndi += stepcoli
        coldmndj += stepcolj
        starti, startj = (starti+incri, startj + incrj)
        row, col = starti, startj

    
    return (resBoard, lx, ly, lastIndex)

# Identify the location of the starting positon
def getStartingPosition(board):
    for i, row in enumerate(board):
        for j, col in enumerate(row):
            if col == 'L':
                return (i,j)
    raise ValueError("Board does not have L!")

# Calculates the partial sum
def constructSumBoard(board, resBoard):
    size = len(board)
    midway = size/2
    coldmndi, coldmndj = midway, midway+1
    stepcoli, stepcolj = -1, 1

    for x in xrange(size):
        resrow = resBoard[x]
        for y in xrange(coldmndi, size, 2):
            resrow[y] = board[x][y]
            if (x-1) >= 0 and (y-1) >= 0: resrow[y] = (resrow[y] + resBoard[x-1][y-1]) % mod
            if (x-2) >= 0: resrow[y] = (resrow[y] + resBoard[x-2][y]) % mod
            if (y-2) >= 0: resrow[y] = (resrow[y] + resBoard[x][y-2]) % mod
            if (x-2) >= 0 and (y-2) >= 0:
                resrow[y] = (resrow[y] - resBoard[x-2][y-2]) % mod
        if x >= midway:
            coldmndi = 0 if coldmndi == 1 else 1


# old method
def constructSumBoardOld(board):
    res = []
    for i, row in enumerate(board):
        resrow = []
        for j, col in enumerate(row):
            resrow.append(col)
            if j > 0: resrow[j] = (resrow[j] + resrow[j-1]) % mod
            if i > 0: resrow[j] = (resrow[j] + res[i-1][j]) % mod
            if j > 0 and i > 0: resrow[j] = (resrow[j] - res[i-1][j-1]) % mod
        res.append(resrow)
    return res

# Recalculate the board values based on the partial sums
def recalculateResultBoard(origboard, sumboard, S, lastIndex, resBoard):
    prevBoard = copy.deepcopy(resBoard)
    size = len(origboard)
    midway = size/2
    coldmndi, coldmndj = midway, midway+1
    stepcoli, stepcolj = -1, 1

    for x in xrange(size):
        resrow = resBoard[x]
        for y in xrange(coldmndi, coldmndj, 2):
            if origboard[x][y] == 1:
                resrow[y] = sumboard[min(size-1,x+S)][min(size-1, y+S)]
                if x-S-1 >= 0:
                    resrow[y] = (resrow[y] - sumboard[x-S-1][min(size-1, y+S-1)])%mod
                if x-S-2 >= 0:
                    resrow[y] = (resrow[y] - prevBoard[x-S-2][min(size-1, y+S)])%mod
                    
                if y-S-1 >= 0:
                    resrow[y] = (resrow[y] - sumboard[min(size-1, x+S-1)][y-S-1])%mod
                if y-S-2 >= 0:
                    resrow[y] = (resrow[y] - prevBoard[min(size-1, x+S)][y-S-2])%mod

                if x-S-1 >= 0 and y-S-1 >= 0:
                    resrow[y] = (resrow[y] + sumboard[x-S-1][y-S-1])%mod

        if coldmndi <= 0:
            stepcoli, stepcolj = (0, stepcoli)
        coldmndi += stepcoli
        coldmndj += stepcolj

# This is DP solution. The idea is to calculate the number of ways to move from
# the every location at each step. For step 1 it would be the total number of
# positions that can be reached from that location given S. For step 2 it would
# be the sum of these values at every location and so on.
# To do this easily for M steps, at every step we need to calculate the partial
# sum just like we do for the classic maxsum DP problem. Using these values, we
# can calculate the total number of ways at each step in O(N^2) ways. This gives
# us the total complexity at O(MN^2).
# Note that to do this partial sums easily, the board must be rotated to a 
# diamond shape so that our calculation is easy.
def fairyChess(N,S,M,board):
    (origboard, Lx, Ly, lastIndex) = initResultBoard(board)
    resBoard = copy.deepcopy(origboard)
    sumBoard = copy.deepcopy(origboard)
    for step in xrange(M):
        constructSumBoard(resBoard, sumBoard)
        temp = constructSumBoardOld(resBoard)
        for x, rowvals in enumerate(temp):
            #print rowvals
            print sumBoard[x]
                
        #print "Sumboard:", sumBoard
        recalculateResultBoard(origboard, sumBoard, S, lastIndex, resBoard)
        #print "Resboard:", resBoard

    return resBoard[Lx][Ly]
    
def main():
    T = int(raw_input())
    for test in xrange(T):
        N,M,S = tuple( [ int(x) for x in raw_input().split() ] )
        board = []
        for row in xrange(N):
            rowstring = raw_input().strip()
            board.append(rowstring)
        print fairyChess(N,S,M,board)

if __name__ == "__main__":
    main()
